src/wbase/FileChannelShared.cc

```diff
@@ -0,0 +1,340 @@
+    // Operation stats. Note that "buffer fill time" included the amount
+    // of time needed to write the result set to disk.
+    util::Timer transmitT;
+    transmitT.start();
+
+    double bufferFillSecs = 0.0;
+    int bytesTransmitted = 0;
+    int rowsTransmitted = 0;
+
+    // Keep reading rows and converting those into messages while any
+    // are still left in the result set. The row processing method
+    // will write rows into the output file. The final "summary" message
+    // will be sant back to Czar after processing the very last set of rows
+    // of the last task of a request.
+    bool erred = false;
+    bool hasMoreRows = true;
+
+    lock_guard<mutex> const tMtxLock(tMtx);
+
+    while (hasMoreRows && !cancelled) {
+        util::Timer bufferFillT;
+        bufferFillT.start();
+
+        // Initialize transmitData, if needed.
+        initTransmit(tMtxLock, *task);
+
+        // Transwer rows from a result set into the data buffer. Note that tSize
+        // is set by fillRows. A value of this variable is presently not used by
+        // the code.
+        size_t tSize = 0;
+        hasMoreRows = !transmitData->fillRows(mResult, tSize);
+
+        // Serialize the content of the data buffer into the Protobuf data message
+        // that will be writen into the output file.
+        transmitData->buildDataMsg(*task, multiErr);
+        _writeToFile(tMtxLock, task, transmitData->dataMsg());
+
+        bufferFillT.stop();
+        bufferFillSecs += bufferFillT.getElapsed();
+
+        int const bytes = transmitData->getResultSize();
+        int const rows = transmitData->getResultRowCount();
+        bytesTransmitted += bytes;
+        rowsTransmitted += rows;
+        _rowcount += rows;
+        _transmitsize += bytes;
+
+        // If no more rows are left in the task's result set then we need to check
+        // if this is last task in a logical group of ones created for processing
+        // the current request (note that certain classes of requests may require
+        // more than one task for processing).
+        if (!hasMoreRows && transmitTaskLast()) {
+            // Make sure the file is sync to disk before notifying Czar.
+            _file.flush();
+            _file.close();
+
+            // Only the last ("summary") message w/o any rows is sent to Czar to notify
+            // the about completion of the request.
+            transmitData->prepareResponse(*task, _rowcount, _transmitsize);
+            bool const lastIn = true;
+            if (!prepTransmit(tMtxLock, task, cancelled, lastIn)) {
+                LOGS(_log, LOG_LVL_ERROR, "Could not transmit the summary message to Czar.");
+                erred = true;
+                break;
+            }
+        } else {
+            // Scrap the transmit buffer to be ready for processing the next set of rows
+            // of the current or the next task of the request.
+            transmitData.reset();
```

Comment:
This is needed because the object now (unlike the original SSI-only) has two purposes - accumulating rows of the partial result sets of tasks to be flushed into a file, and preparing the final response to be sent back to `Czar`. If the object isn't reset then it's possible that the last message (sent to `Czar`) would be "contaminated" with the data. So, the reset is the safest action here to ensure the object is always in a clean state before being used.